syntax = "proto3";

// Not necessary for Python but should still be declared to avoid name collisions
// in the Protocol Buffers namespace and non-Python languages
package protoblog;

enum ParameterKind {
    POSITIONAL_ONLY = 0;
    POSITIONAL_OR_KEYWORD = 1;
    KEYWORD_ONLY = 2;
    VAR_KEYWORD = 3;
    VAR_POSITIONAL = 4;
}

enum TypeKind {
    INSTANCE = 0;
    UNION = 1;
    TYPE = 2;
    TUPLE = 3;
    TYPE_VAR = 4;
    ANY = 5;
    NONE = 6;
    TYPE_ALIAS = 7;
    CALLABLE = 8;
    LITERAL = 9;
    UNINHABITED = 10;
    UNBOUND = 11;
    TYPED_DICT = 12;
}

message Type {
    string pretty_printed_name = 1;
    TypeKind kind = 2;
    repeated Type args = 3;
    optional string simple_name = 4;
}

message ParameterSymbol {
        string name = 1;
        ParameterKind kind = 2;
        optional Type type_annotation = 3;
        bool has_default = 4;
}

message FunctionSymbol {
    string name = 1;
    string fully_qualified_name = 2;
    optional Type return_annotation = 4;
    repeated ParameterSymbol parameters = 5;
    bool has_decorators = 6;
    repeated string resolved_decorator_names = 7;
    bool is_abstract = 8;
    bool is_async_generator = 9;
    bool is_awaitable_coroutine = 10;
    bool is_coroutine = 11;
    bool is_final = 12;
    bool is_overload = 13;
    bool is_property = 14;
    bool is_static = 15;
    bool is_class_method = 16;
}

message OverloadedFunctionSymbol {
    string name = 1;
    string fullname = 2;
    bool is_final = 3;
    bool is_property = 4;
    bool is_static = 5;
    repeated FunctionSymbol definitions = 6;
}

message ClassSymbol {
    string name = 1;
    string fully_qualified_name = 2;
    repeated string super_classes = 4;
    repeated string mro = 5;
    repeated FunctionSymbol methods = 6;
    repeated OverloadedFunctionSymbol overloaded_methods = 7;
    bool has_decorators = 8;
    bool has_metaclass = 9;
    bool is_enum = 10;
    bool is_generic = 11;
    bool is_named_tuple = 12;
    bool is_protocol = 13;
    optional string metaclass_name = 14;
    optional Type metaclass_type = 15;
}

message ModuleSymbol {
    string name = 1;
    string fully_qualified_name = 2;
    repeated ClassSymbol classes = 3;
    repeated FunctionSymbol functions = 4;
    repeated OverloadedFunctionSymbol overloaded_functions = 5;
}
